README.md
# Cloudability - Money

## Contributing

When contributing, please make sure to update the CHANGELOG and AUTHORS files
when you submit your pull request. Upon merging of your first pull request,
you will be given commit access to the repository.

## Introduction

This library aids one in handling money and different currencies.

### Features

- Provides a `Money` class which encapsulates all information about an certain
  amount of money, such as its value and its currency.
- Provides a `Money::Currency` class which encapsulates all information about
  a monetary unit.
- Represents monetary values as integers, in cents. This avoids floating point
  rounding errors.
- Represents currency as `Money::Currency` instances providing an high level of
  flexibility.
- Provides APIs for exchanging money from one currency to another.
- Has the ability to parse a money and currency strings
  into the corresponding Money/Currency object.

### Resources

- [Git Repository](http://github.com/MrJoy/money)

### Notes

- Your app must use UTF-8 to function with this library. There are a
  number of non-ASCII currency attributes.
- This app requires Ruby 1.9 and JSON. If you're using JRuby < 1.7.0
  you'll need to add `gem "json"` to your Gemfile or similar.

## Downloading

Install stable releases with the following command:

    gem install money

The development version (hosted on Github) can be installed with:

    git clone git://github.com/MrJoy/money.git
    cd money
    rake install

## Usage

``` ruby
require 'money'

# 10.00 USD
money = Money.new(1000, "USD")
money.cents     #=> 1000
money.currency  #=> Currency.new("USD")

# Comparisons
Money.new(1000, "USD") == Money.new(1000, "USD")   #=> true
Money.new(1000, "USD") == Money.new(100, "USD")    #=> false
Money.new(1000, "USD") == Money.new(1000, "EUR")   #=> false
Money.new(1000, "USD") != Money.new(1000, "EUR")   #=> true

# Arithmetic
Money.new(1000, "USD") + Money.new(500, "USD") == Money.new(1500, "USD")
Money.new(1000, "USD") - Money.new(200, "USD") == Money.new(800, "USD")
Money.new(1000, "USD") / 5                     == Money.new(200, "USD")
Money.new(1000, "USD") * 5                     == Money.new(5000, "USD")

# Assumptive Currencies
Money.assume_from_symbol = true
Money.parse("$100") == Money.new(10000, "USD")
Money.parse("€100") == Money.new(10000, "EUR")
Money.parse("£100") == Money.new(10000, "GBP")

# Currency conversions
some_code_to_setup_exchange_rates
Money.new(1000, "USD").exchange_to("EUR") == Money.new(some_value, "EUR")
```

## Currency

Currencies are consistently represented as instances of `Money::Currency`.
The most part of `Money` APIs allows you to supply either a `String` or a
`Money::Currency`.

``` ruby
Money.new(1000, "USD") == Money.new(1000, Currency.new("USD"))
Money.new(1000, "EUR").currency == Currency.new("EUR")
```

A `Money::Currency` instance holds all the information about the currency,
including the currency symbol, name and much more.

``` ruby
currency = Money.new(1000, "USD").currency
currency.iso_code #=> "USD"
currency.name     #=> "United States Dollar"
```

To define a new `Money::Currency` use `Money::Currency.register` as shown
below.

``` ruby
curr = {
  :priority        => 1,
  :iso_code        => "USD",
  :iso_numeric     => "840",
  :name            => "United States Dollar",
  :symbol          => "$",
  :subunit         => "Cent"
  :subunit_to_unit => 100,
  :separator       => ".",
  :delimiter       => ","
}

Money::Currency.register(curr)
```

The pre-defined set of attributes includes:

- `:priority` a numerical value you can use to sort/group the currency list
- `:iso_code` the international 3-letter code as defined by the ISO 4217 standard
- `:iso_numeric` the international 3-digit code as defined by the ISO 4217 standard
- `:name` the currency name
- `:symbol` the currency symbol (UTF-8 encoded)
- `:subunit` the name of the fractional monetary unit
- `:subunit_to_unit` the proportion between the unit and the subunit
- `:separator` character between the whole and fraction amounts
- `:delimiter` character between each thousands place

All attributes are optional. Some attributes, such as `:symbol`, are used by
the Money class to print out a representation of the object. Other attributes,
such as `:name` or `:priority`, exist to provide a basic API you can take
advantage of to build your application.

### :priority

The priority attribute is an arbitrary numerical value you can assign to the
`Money::Currency` and use in sorting/grouping operation.

For instance, let's assume your Rails application needs to render a currency
selector like the one available
[here](http://finance.yahoo.com/currency-converter/). You can create a couple of
custom methods to return the list of major currencies and all currencies as
follows:

``` ruby
# Returns an array of currency id where
# priority < 10
def major_currencies(hash)
  hash.inject([]) do |array, (id, attributes)|
    priority = attributes[:priority]
    if priority && priority < 10
      array[priority] ||= []
      array[priority] << id
    end
    array
  end.compact.flatten
end

# Returns an array of all currency id
def all_currencies(hash)
  hash.keys
end

major_currencies(Money::Currency.table)
# => [ :usd, :eur, :bgp, :cad ]

all_currencies(Money::Currency.table)
# => [ :aed, :afn, all, ... ]
```

### Default Currency

By default `Money` defaults to USD as its currency. This can be overwritten
using:

``` ruby
Money.default_currency = Money::Currency.new("CAD")
```

If you use Rails, then `environment.rb` is a very good place to put this.

## Currency Exchange

Exchanging money is performed through an exchange bank object. The default
exchange bank object requires one to manually specify the exchange rate. Here's
an example of how it works:

``` ruby
Money.add_rate("USD", "CAD", 1.24515)
Money.add_rate("CAD", "USD", 0.803115)

Money.us_dollar(100).exchange_to("CAD")  # => Money.new(124, "CAD")
Money.ca_dollar(100).exchange_to("USD")  # => Money.new(80, "USD")
```

Comparison and arithmetic operations work as expected:

``` ruby
Money.new(1000, "USD") <=> Money.new(900, "USD")   # => 1; 9.00 USD is smaller
Money.new(1000, "EUR") + Money.new(10, "EUR") == Money.new(1010, "EUR")

Money.add_rate("USD", "EUR", 0.5)
Money.new(1000, "EUR") + Money.new(1000, "USD") == Money.new(1500, "EUR")
```

There is nothing stopping you from creating bank objects which scrapes
[XE](http://www.xe.com) for the current rates or just returns `rand(2)`:

``` ruby
Money.default_bank = ExchangeBankWhichScrapesXeDotCom.new
```

### Implementations

The following is a list of Money.gem compatible currency exchange rate
implementations.

- [eu_central_bank](http://github.com/RubyMoney/eu_central_bank)
- [google_currency](http://github.com/RubyMoney/google_currency)
- [nordea](https://github.com/k33l0r/nordea)
- [nbrb_currency](https://github.com/slbug/nbrb_currency)
- [money-open-exchange-rates](https://github.com/spk/money-open-exchange-rates)
- [money-historical-bank](https://github.com/atwam/money-historical-bank)

## Ruby on Rails

To integrate money in a Rails application use [money-rails](http://github.com/RubyMoney/money-rails)
gem or follow the instructions below.

Define accessor methods to let Active Record deal with embedding the money
object in your models. The following example requires 2 columns:

``` ruby
:price_cents, :integer, :default => 0, :null => false
:price_currency, :string
```

Then in your model file:

``` ruby
def price
  Money.new price_cents || 0, price_currency || Money.default_currency
end

def price=(value)
  Money.parse(value).tap do |price|
    write_attribute :price_cents,    price.cents
    write_attribute :price_currency, price.currency_as_string
  end
end
```


CHANGELOG.md
# Changelog

## master

NOTICE: This change breaks the magic hand-waving about commas-vs-periods as
thousands separators vs. decimal separators.  Now, the currency's configuration
will always be respected in this regard.  This is necessary to properly handle
high-precision values.

- Add support for retaining high-precision versions of money values.
- Reformat changelog to actually be Markdown.
- Add a numeric ID for currencies that can fit in an unsigned 1-byte int, handy
  for dimension columns in ROLAP tables.  Addded the `assign_ids` rake task to
  help keep things up to date.

## upstream

- Calling Bank::Base.instance doesn't make Bank::VariableExchange.instance
  return Bank::Base.instance anymore (semaperepelitsa)
- Update Turkmenistan manat from TMM to TMT currency (GH-181). [Thanks @Exoth]
- Moved ZWD Zimbabwean dollars to currency_bc.json, also added there ZWL, ZWN, and ZWR Zimbabwean dollars (GH-184).

## 5.1.0

- Switch to multi_json gem (GH-170)
- Fix "warning: ambiguous first argument..." (GH-166)
- Update dependencies to latest and greatest (GH-172)
- TravisBot is now watching Pull Request!!! (GH-171)
- Minor code cleaning

## 5.0.0

- Minor bugfix - incorrect use of character range resulted in
  botched results for Money::Parsing#extract_cents (GH-162)
- Money::Currency::TABLE removed. Use Money::Currency.register to add
  additional currencies (GH-143)
- Fix rounding error in Numeric.to_money (GH-145)
- Allow on-the-fly calculation of decimal places if not known already
  (GH-146,GH-147,GH-148)
- Move Euro symbol ahead of amount (GH-151)
- Changed settings for Polish currency (GH-152)
- Fall back to symbol if there is no html_entity present (GH-153)
- Optionally Allow parsing of money values prefixed by symbols in key
  currencies (GH-155)
- Fix bug where rates exported to a file from VariableExchange leave the File
  object open (GH-154)
- Added Money#positive? and Money#negative? methods (GH-157)
- Fix format function output for custom currencies (GH-156)
- Fix parsing of strings with 3 decimal digits (GH-158)
- Updated development dependencies
- Said goodbye to RubyForge


## 4.0.2

- Money.to_money now understands a currency option (GH-121)
- Added Money#-@ method to change object polarity (GH-122)
- Added Money#symbol_or_iso_code utility method (GH-128)
- Money.parse now understands trailing - as negative inputs (GH-133)
- Money::Currency.new now validates input to avoid memory leaks (GH-137)

- Forced UTF-8 encoding on currency JSON (GH-117)
- Fixed position of Philippine peso sign (GH-124)
- Fixed position of Danish currency sign (GH-127)


## 4.0.1

- Add missing config dir.


## 4.0.0

The big change this release is moving the currency information into a JSON
file. This should make it easier for users to add and test things. Thanks to
Steve Morris for working on this during his Mendicant University course.

In addition to this big change there were well over a dozen other minor
changes.

### Features

- a new exchange bank nordea has been added to the README. (k33l0r)
- a new exchange bank nbrb_currency has been added to the README. (slbug)
- update Rake tasks
- See our CI status in the README
- Add syntax highlighting to the README (phlipper)
- Remove minor unit from YEN (pwim)
- Format YEN (pwim)
- Update README for `_as_string` (mjankowski)
- Update Lebanon currency (kaleemullah)
- Update Polish złoty (holek)
- Move currency information into JSON storage! (stevemorris)
- Add ISO4217 Numeric codes (alovak)

### Bugfixes

- EEK currency is no longer used, kept for BC ([#issue/110](http://github.com/RubyMoney/money/issues/110))
- Lithuanian Litas symbol position fixed (laurynas)
- Fixed README typos (phlipper)
- Fixed README typos (pwim)
- Fix specs (alovak)


## 3.7.1

### Bugfixes

- Add encoding indicator to top of Gemspec


## 3.7.0

### Features

- add Money#to_d (thanks Andrew White)
- Add Money.use_i18n, this allows you to enable/disable i18n from being used,
  even if it's required in your app.


## 3.6.2

### Features

- i18n enhancements (thanks eloyesp [link](https://github.com/RubyMoney/money/commit/b2cab76c78ae04f40251fa20c4ab18faa968dc53))
- README updates (thanks pconnor)
- Break into modules
- Added `:no_cents_if_whole` format option
- Update HKD from Ho to Cent
- Performance improvements (thanks weppos)
- Added Symbol#to_currency
- Added Gemfile for development
- Updated HUF currency to use `symbol_first => false`
- Updated "Turkish New Lira" to "Turkish Lira"


## 3.6.1

### Bugfixes

- Floating point comparison needs to Epsilon aware (thanks Tobias Luetke)
- reimplement fix for #issue/43, enable Marshal.(load/dump)


## 3.6.0

### Features

- Add a symbol position option for Money#format (thanks Romain, Gil and
  Julien)
- Updated CNY to use "Fen" and subunit_to_unit of 100
- Updates to work with gem-testers.org

### Bugfixes

- Fixed issue with #format(:no_cents => true) (thanks Romain & Julien)


## 3.5.5

### Features

- Enhancements to Money::Currency (thanks Matthew McEachen)
  - Replace delimiter with thousands_separator
  - Replace separator with decimal_mark
  - Added symbol_first and html_entity
- Added allocation algorithm for fair(ish) splitting of money between parties
  without loosing pennies (thanks Tobias Luetke)

### Bugfixes

- Always store cents as an Integer (thanks Abhay Kumar)
- Fixed TypeError in rate exchange (thanks Christian Billen)
- Cleanup #parse (thanks Tom Lianza)


## 3.5.4

### Features

- Added Currency#decimal_places.

### Bugfixes

- Fixed error with Money#to_s error with negative amounts that are only cents.


## 3.5.3

### Bugfixes

- Fixed an error in #to_s when cents is negative


## 3.5.2

### Bugfixes

- Fixed an error in #to_s which appended extra 0s incorrectly


## 3.5.1

### Bugfixes

- Removed erroneous require.


## 3.5.0

### Features

- Updated to RSpec2.
- Use i18n to lookup separator and delimiter signs.
- Removed all deprecated methods up to v3.5.0, including the following:

spec/currency_spec.rb
# encoding: utf-8

require "spec_helper"

describe Money::Currency do

  FOO = '{ "priority": 1, "iso_code": "FOO", "iso_numeric": "840", "name": "United States Dollar", "symbol": "$", "subunit": "Cent", "subunit_to_unit": 450, "symbol_first": true, "html_entity": "$", "decimal_mark": ".", "thousands_separator": "," }'

  describe ".find" do
    it "returns currency matching given id" do
      Money::Currency.register(JSON.parse(FOO, :symbolize_names => true))

      expected = Money::Currency.new(:foo)
      Money::Currency.find(:foo).should  == expected
      Money::Currency.find(:FOO).should  == expected
      Money::Currency.find("foo").should == expected
      Money::Currency.find("FOO").should == expected
    end

    it "returns nil unless currency matching given id" do
      Money::Currency.find("ZZZ").should be_nil
    end
  end

  describe ".wrap" do
    it "returns nil if object is nil" do
      Money::Currency.wrap(nil).should == nil
      Money::Currency.wrap(Money::Currency.new(:usd)).should == Money::Currency.new(:usd)
      Money::Currency.wrap(:usd).should == Money::Currency.new(:usd)
    end
  end


  describe "#initialize" do
    it "lookups data from loaded config" do
      currency = Money::Currency.new("USD")
      currency.key.should                 == :usd
      currency.priority.should            == 1
      currency.iso_code.should            == "USD"
      currency.iso_numeric.should         == "840"
      currency.name.should                == "United States Dollar"
      currency.decimal_mark.should        == "."
      currency.separator.should           == "."
      currency.thousands_separator.should == ","
      currency.delimiter.should           == ","
    end

    it "raises UnknownMoney::Currency with unknown currency" do
      lambda { Money::Currency.new("xxx") }.should raise_error(Money::Currency::UnknownCurrency, /xxx/)
    end
  end

  describe "#<=>" do
    it "compares objects by priority" do
      Money::Currency.new(:cad).should > Money::Currency.new(:usd)
      Money::Currency.new(:usd).should < Money::Currency.new(:eur)
    end
  end

  describe "#==" do
    it "returns true if self === other" do
      currency = Money::Currency.new(:eur)
      currency.should == currency
    end

    it "returns true if the id is equal" do
      Money::Currency.new(:eur).should     == Money::Currency.new(:eur)
      Money::Currency.new(:eur).should_not == Money::Currency.new(:usd)
    end
  end

  describe "#eql?" do
    it "returns true if #== returns true" do
      Money::Currency.new(:eur).eql?(Money::Currency.new(:eur)).should be true
      Money::Currency.new(:eur).eql?(Money::Currency.new(:usd)).should be false
    end
  end

  describe "#hash" do
    it "returns the same value for equal objects" do
      Money::Currency.new(:eur).hash.should == Money::Currency.new(:eur).hash
      Money::Currency.new(:eur).hash.should_not == Money::Currency.new(:usd).hash
    end

    it "can be used to return the intersection of Currency object arrays" do
      intersection = [Money::Currency.new(:eur), Money::Currency.new(:usd)] & [Money::Currency.new(:eur)]
      intersection.should == [Money::Currency.new(:eur)]
    end
  end

  describe "#inspect" do
    it "works as documented" do
      Money::Currency.new(:usd).inspect.should ==
          %Q{#<Money::Currency id: 146, key: usd, priority: 1, symbol_first: true, thousands_separator: ,, html_entity: $, decimal_mark: ., name: United States Dollar, symbol: $, subunit_to_unit: 100, iso_code: USD, iso_numeric: 840, subunit: Cent>}
    end
  end

  describe "#to_s" do
    it "works as documented" do
      Money::Currency.new(:usd).to_s.should == "USD"
      Money::Currency.new(:eur).to_s.should == "EUR"
    end
  end

  describe "#to_currency" do
    it "works as documented" do
      usd = Money::Currency.new(:usd)
      usd.to_currency.should == usd
    end

    it "doesn't create new symbols indefinitely" do
      expect { Money::Currency.new("bogus") }.to raise_exception(Money::Currency::UnknownCurrency)
      Symbol.all_symbols.map{|s| s.to_s}.should_not include("bogus")
    end
  end

  describe "#code" do
    it "works as documented" do
      Money::Currency.new(:usd).code.should == "$"
      Money::Currency.new(:azn).code.should == "AZN"
    end
  end

  describe "#decimal_places" do
    it "proper places for known currency" do
      Money::Currency.new(:mro).decimal_places == 1
      Money::Currency.new(:usd).decimal_places == 2
    end

    it "proper places for custom currency" do
      Money::Currency.register(JSON.parse(FOO, :symbolize_names => true))
      Money::Currency.new(:foo).decimal_places == 3
    end
  end
end


lib/money/core_extensions.rb
# Open +Numeric+ to add new method.
class Numeric

  # Converts this numeric into a +Money+ object in the given +currency+.
  #
  # @param [Currency, String, Symbol] currency
  #   The currency to set the resulting +Money+ object to.
  #
  # @return [Money]
  #
  # @example
  #   100.to_money                   #=> #<Money @cents=10000>
  #   100.37.to_money                #=> #<Money @cents=10037>
  #   BigDecimal.new('100').to_money #=> #<Money @cents=10000>
  #
  # @see Money.from_non_string
  #
  def to_money(currency = nil)
    Money.from_non_string(self, currency || Money.default_currency)
  end

end

# Open +String+ to add new methods.
class String

  # Parses the current string and converts it to a +Money+ object.
  # Excess characters will be discarded.
  #
  # @param [Currency, String, Symbol] currency
  #   The currency to set the resulting +Money+ object to.
  #
  # @return [Money]
  #
  # @example
  #   '100'.to_money                #=> #<Money @cents=10000>
  #   '100.37'.to_money             #=> #<Money @cents=10037>
  #   '100 USD'.to_money            #=> #<Money @cents=10000, @currency=#<Money::Currency id: usd>>
  #   'USD 100'.to_money            #=> #<Money @cents=10000, @currency=#<Money::Currency id: usd>>
  #   '$100 USD'.to_money           #=> #<Money @cents=10000, @currency=#<Money::Currency id: usd>>
  #   'hello 2000 world'.to_money   #=> #<Money @cents=200000 @currency=#<Money::Currency id: usd>>
  #
  # @see Money.from_string
  #
  def to_money(currency = nil)
    Money.parse(self, currency)
  end

  # Converts the current string into a +Currency+ object.
  #
  # @return [Money::Currency]
  #
  # @raise [Money::Currency::UnknownCurrency]
  #   If this String reference an unknown currency.
  #
  # @example
  #   "USD".to_currency #=> #<Money::Currency id: usd>
  #
  def to_currency
    Money::Currency.new(self)
  end

end

# Open +Symbol+ to add new methods.
class Symbol

  # Converts the current symbol into a +Currency+ object.
  #
  # @return [Money::Currency]
  #
  # @raise [Money::Currency::UnknownCurrency]
  #   If this String reference an unknown currency.
  #
  # @example
  #   :ars.to_currency #=> #<Money::Currency id: ars>
  #
  def to_currency
    Money::Currency.new(self)
  end

end


lib/money/currency_loader.rb
module CurrencyLoader
  extend self

  DATA_PATH = File.expand_path("../../../config", __FILE__)

  # Loads and returns the currencies stored in JSON files in the config directory.
  #
  # @return [Hash]
  def load_currencies(silence_missing_ids_error = false)
    currencies = parse_currency_file("currency.json")
    currencies.merge! parse_currency_file("currency_bc.json")
    currencies.each do |key, currency|
      if(!currency.has_key?(:key))
        currency[:key] = key
      else
        raise Exception.new("Currency already has key (#{currency[:key]}) but we want to give it a new one (#{key})!")
      end
    end

    begin
      parse_currency_file("currency_ids.json").
        each_pair { |key, id| currencies[key][:id] = id }
      missing_ids = currencies.values.select { |currency| !currency.has_key?(:id) }

      if(missing_ids.count > 0)
        missing_keys = missing_ids.map { |currency| currency[:iso_code] }.join(', ')
        raise Exception.new("The following currencies are missing short IDs.  Go run 'rake assign_ids' to fix it: #{missing_keys}")
      end
    rescue Exception => e
      unless(silence_missing_ids_error)
        raise e
      end
    end

    return currencies
  end

  private

  def parse_currency_file(filename)
    json = File.read("#{DATA_PATH}/#{filename}")
    json.force_encoding(::Encoding::UTF_8) if defined?(::Encoding)
    JSON.parse(json, :symbolize_names => true)
  end
end


lib/money/money/parsing.rb
#encoding: utf-8

class Money
  module Parsing
    def self.included(base)
      base.extend ClassMethods
    end

    module ClassMethods
      # Parses the current string and converts it to a +Money+ object.
      # Excess characters will be discarded.
      #
      # @param [String, #to_s] input The input to parse.
      # @param [Currency, String, Symbol] currency The currency format.
      #   The currency to set the resulting +Money+ object to.
      #
      # @return [Money]
      #
      # @raise [ArgumentError] If any +currency+ is supplied and
      #   given value doesn't match the one extracted from
      #   the +input+ string.
      #
      # @example
      #   '100'.to_money                #=> #<Money @cents=10000>
      #   '100.37'.to_money             #=> #<Money @cents=10037>
      #   '100 USD'.to_money            #=> #<Money @cents=10000, @currency=#<Money::Currency id: usd>>
      #   'USD 100'.to_money            #=> #<Money @cents=10000, @currency=#<Money::Currency id: usd>>
      #   '$100 USD'.to_money           #=> #<Money @cents=10000, @currency=#<Money::Currency id: usd>>
      #   'hello 2000 world'.to_money   #=> #<Money @cents=200000 @currency=#<Money::Currency id: usd>>
      #
      # @example Mismatching currencies
      #   'USD 2000'.to_money("EUR")    #=> ArgumentError
      #
      # @see Money.from_string
      #
      def parse(input, currency = nil)
        i = input.to_s.strip

        # raise Money::Currency.table.collect{|c| c[1][:symbol]}.inspect

        # Check the first character for a currency symbol, alternatively get it
        # from the stated currency string
        c = if Money.assume_from_symbol && i =~ /^(\$|€|£)/
          case i
          when /^$/ then "USD"
          when /^€/ then "EUR"
          when /^£/ then "GBP"
          end
        else
          i[/[A-Z]{2,3}/]
        end

        # check that currency passed and embedded currency are the same,
        # and negotiate the final currency
        if currency.nil? and c.nil?
          currency = Money.default_currency
        elsif currency.nil?
          currency = c
        elsif c.nil?
          currency = currency
        elsif currency != c
          # TODO: ParseError
          raise ArgumentError, "Mismatching Currencies"
        end
        currency = Money::Currency.wrap(currency)

        cents = extract_cents(i, currency)
        new(cents, currency)
      end

      # Converts a String into a Money object treating the +value+
      # as dollars and converting them to the corresponding cents value,
      # according to +currency+ subunit property,
      # before instantiating the Money object.
      #
      # Behind the scenes, this method relies on {Money.from_bigdecimal}
      # to avoid problems with string-to-numeric conversion.
      #
      # @param [String, #to_s] value The money amount, in dollars.
      # @param [Currency, String, Symbol] currency
      #   The currency to set the resulting +Money+ object to.
      #
      # @return [Money]
      #
      # @example
      #   Money.from_string("100")
      #   #=> #<Money @cents=10000 @currency="USD">
      #   Money.from_string("100", "USD")
      #   #=> #<Money @cents=10000 @currency="USD">
      #   Money.from_string("100", "EUR")
      #   #=> #<Money @cents=10000 @currency="EUR">
      #   Money.from_string("100", "BHD")
      #   #=> #<Money @cents=100 @currency="BHD">
      #
      # @see String#to_money
      # @see Money.parse
      #
      def from_string(value, currency = Money.default_currency)
        from_non_string(BigDecimal.new(value.to_s), currency)
      end

      # Converts a numeric type into a Money object treating the +value+
      # as dollars and converting them to the corresponding cents value,
      # according to +currency+ subunit property,
      # before instantiating the Money object.
      #
      # @param [Fixnum, Float, or BigDecimal] value The money amount, in dollars.
      # @param [Currency, String, Symbol] currency The currency format.
      #
      # @return [Money]
      #
      # @example
      #   Money.from_non_string(100)
      #   #=> #<Money @cents=10000 @currency="USD">
      #   Money.from_non_string(100, "USD")
      #   #=> #<Money @cents=10000 @currency="USD">
      #   Money.from_non_string(100, "EUR")
      #   #=> #<Money @cents=10000 @currency="EUR">
      #   Money.from_non_string(100, "BHD")
      #   #=> #<Money @cents=100 @currency="BHD">
      #
      # @see Fixnum#to_money
      # @see Money.from_non_string
      #
      def from_non_string(value, currency = Money.default_currency)
        currency = Money::Currency.wrap(currency)
        amount   = value * currency.subunit_to_unit
        new(amount, currency)
      end

      # Takes a number string and attempts to massage out the number.
      #
      # @param [String] input The string containing a potential number.
      #
      # @return [Integer]
      #
      def extract_cents(input, currency = Money.default_currency)
        # remove anything that's not a number, potential thousands_separator, or minus sign
        num = input.
          gsub(/[^\d\.,\'\-]/, '').
          strip

        # set a boolean flag for if the number is negative or not
        negative = num =~ /^-|-$/ ? true : false

        # decimal mark character
        decimal_char = currency.decimal_mark

        # if negative, remove the minus sign from the number
        # if it's not negative, the hyphen makes the value invalid
        if negative
          num = num.sub(/^-|-$/, '')
        end

        raise ArgumentError, "Invalid currency amount (hyphen)" if num.include?('-')

        num = num.gsub(currency.delimiter, '').gsub(currency.separator, '.')
        digits = currency.subunit_to_unit.to_s.length - 1
        # TODO: This is an ugly hack to address the MGA tests failing, but
        # TODO: honestly, I'm not sure the behavior is 'correct' when we go to
        # TODO: more than a digit of precision for it.
        digits = 1 if digits < 1
        major = num.to_i
        minor = (BigDecimal.new(num) - major) * (10 ** digits)
        cents = (major * currency.subunit_to_unit) + minor

          # gather all decimal_marks within the result number
          used_decimal_marks = num.scan /[^\d]/

          # determine the number of unique decimal_marks within the number
          #
          # e.g.
          # $1,234,567.89 would return 2 (, and .)
          # $125,00 would return 1
          # $199 would return 0
          # $1 234,567.89 would raise an error (decimal_marks are space, comma, and period)
          case used_decimal_marks.uniq.length
            # no decimal_mark or thousands_separator; major (dollars) is the number, and minor (cents) is 0
          when 0 then major, minor = num, 0

            # two decimal_marks, so we know the last item in this array is the
            # major/minor thousands_separator and the rest are decimal_marks
          when 2
            decimal_mark, thousands_separator = used_decimal_marks.uniq
            # remove all decimal_marks, split on the thousands_separator
            major, minor = num.gsub(decimal_mark, '').split(thousands_separator)
            min = 0 unless min
          when 1
            # we can't determine if the comma or period is supposed to be a decimal_mark or a thousands_separator
            # e.g.
            # 1,00 - comma is a thousands_separator
            # 1.000 - period is a thousands_separator
            # 1,000 - comma is a decimal_mark
            # 1,000,000 - comma is a decimal_mark
            # 10000,00 - comma is a thousands_separator
            # 1000,000 - comma is a thousands_separator

            # assign first decimal_mark for reusability
            decimal_mark = used_decimal_marks.first

            # When we have identified the decimal mark character
            if decimal_char == decimal_mark
              major, minor = num.split(decimal_char)

            else
              # decimal_mark is used as a decimal_mark when there are multiple instances, always
              if num.scan(decimal_mark).length > 1 # multiple matches; treat as decimal_mark
                major, minor = num.gsub(decimal_mark, ''), 0
              else
                # ex: 1,000 - 1.0000 - 10001.000
                # split number into possible major (dollars) and minor (cents) values
                possible_major, possible_minor = num.split(decimal_mark)
                possible_major ||= "0"
                possible_minor ||= "00"

                # if the minor (cents) length isn't 3, assign major/minor from the possibles
                # e.g.
                #   1,00 => 1.00
                #   1.0000 => 1.00
                #   1.2 => 1.20
                if possible_minor.length != 3 # thousands_separator
                  major, minor = possible_major, possible_minor
                else
                  # minor length is three
                  # let's try to figure out intent of the thousands_separator

                  # the major length is greater than three, which means
                  # the comma or period is used as a thousands_separator
                  # e.g.
                  #   1000,000
                  #   100000,000
                  if possible_major.length > 3
                    major, minor = possible_major, possible_minor
                  else
                    # number is in format ###{sep}### or ##{sep}### or #{sep}###
                    # handle as , is sep, . is thousands_separator
                    if decimal_mark == '.'
                      major, minor = possible_major, possible_minor
                    else
                      major, minor = "#{possible_major}#{possible_minor}", 0
                    end
                  end
                end
              end
            end
          else
            # TODO: ParseError
            raise ArgumentError, "Invalid currency amount"
          end

        # build the string based on major/minor since decimal_mark/thousands_separator have been removed
        # avoiding floating point arithmetic here to ensure accuracy
        cents = (major.to_i * currency.subunit_to_unit)
        # Because of an bug in JRuby, we can't just call #floor
        minor = minor.to_s
        minor = if minor.size < currency.decimal_places
                  (minor + ("0" * currency.decimal_places))[0,currency.decimal_places].to_i
                elsif minor.size > currency.decimal_places
                  if minor[currency.decimal_places,1].to_i >= 5
                    minor[0,currency.decimal_places].to_i+1
                  else
                    minor[0,currency.decimal_places].to_i
                  end
                else
                  minor.to_i
                end
        cents += minor

        # if negative, multiply by -1; otherwise, return positive cents
        cents = negative ? cents * -1 : cents
        return cents
      end
    end
  end
end
